tests/common/test_logger.py

```diff
@@ -229,6 +229,98 @@ def test_telemetry_logger_not_on_by_default(self, mock_save):
                                                 'Microsoft.OSTCExtensions.DummyExtension-1.2.3.4')
         self.assertEqual(0, mock_save.call_count)
 
+    @patch("azurelinuxagent.common.logger.StdoutAppender.write")
+    @patch("azurelinuxagent.common.logger.TelemetryAppender.write")
+    @patch("azurelinuxagent.common.logger.ConsoleAppender.write")
+    @patch("azurelinuxagent.common.logger.FileAppender.write")
+    def test_add_appender(self, mock_file_write, mock_console_write, mock_telem_write, mock_stdout_write):
+        lg = logger.Logger(logger.DEFAULT_LOGGER, "YoloLogger")
+
+        lg.add_appender(logger.AppenderType.FILE, logger.LogLevel.INFO, path=self.log_file)
+        lg.add_appender(logger.AppenderType.TELEMETRY, logger.LogLevel.WARNING, path=add_log_event)
+        lg.add_appender(logger.AppenderType.CONSOLE, logger.LogLevel.WARNING, path="/dev/null")
+        lg.add_appender(logger.AppenderType.STDOUT, logger.LogLevel.WARNING, path=None)
+
+        counter = 0
+        for appender in lg.appenders:
+            if isinstance(appender, logger.FileAppender):
+                counter += 1
+            elif isinstance(appender, logger.TelemetryAppender):
+                counter += 1
+            elif isinstance(appender, logger.ConsoleAppender):
+                counter += 1
+            elif isinstance(appender, logger.StdoutAppender):
+                counter += 1
+
+        # All 4 appenders should have been included.
+        self.assertEquals(4, counter)
+
+        lg.warn("Test Log")
+        self.assertEqual(mock_file_write.call_count, 1)
+        self.assertEqual(mock_console_write.call_count, 1)
+        self.assertEqual(mock_telem_write.call_count, 1)
+        self.assertEqual(mock_stdout_write.call_count, 1)
+
+        lg.info("Test Log")
```

Comment:
Correct me if I'm wrong, but for info, shouldn't only the file_appender be called?

Reply:
Yup. The test is incorrectly checking the write. The write for all the loggers gets called, but the levels are honored in the write method itself.

Eg:
```
    def write(self, level, msg):
        if self.level <= level:
            try:
                with open(self.path, "w") as console:
                    console.write(msg)
            except IOError:
                pass
```

Reply:
Each appender needs its own test to validate the writing of the log message for different levels.

Reply:
Done. Added a new set of tests for each appender.

Reply:
could you add the reply to the original question as a comment in the code? thanks

also for the nested logger test below

Reply:
Done.